### PR TITLE
ENH: 2-D tsa.seasonal_decompose with trend extrapolation

### PR DESCRIPTION
**Pull request by kernc**
fixes https://github.com/statsmodels/statsmodels/issues/2820
closes https://github.com/statsmodels/statsmodels/pull/3684

Includes patch from https://github.com/statsmodels/statsmodels/pull/3684, with tests.

### CODE REVIEW

**Comment by josef-pkt** [comment]
Because I cannot block a gravatar, I blocked the user from the statsmodels organization

**Comment by josef-pkt** [comment]
Thanks for the change in gravatar.

I'm currently busy with other things but try to look at it soon.

File: statsmodels/tsa/seasonal.py [hunk 1]
```diff
@@ -18,7 +18,31 @@ def seasonal_mean(x, freq):
     return np.array([pd_nanmean(x[i::freq]) for i in range(freq)])
 
 
-def seasonal_decompose(x, model="additive", filt=None, freq=None, two_sided=True):
+def _interpolate_trend(trend, npoints):
+    """
+    Replace nan values on trend's end-opoints with least-squares interpolated
+    values wiith regression considering npoints closest defined points.
```
Comment (josef-pkt) [inline statsmodels/tsa/seasonal.py]: two typos end-opoints and wiith

**Comment by josef-pkt** [comment]
Thanks, looks interesting and useful.

I will have to look more closely at the details.
One issue: the plan is to vectorize the function, see #2820. There shouldn't be any problems vectorizing the linear trend estimation, but that needs verification.

(and this makes me think that we could use other methods for trend estimation in this case, 
local linear regression is also a convolution filter with equal spaced time points.
signal.convolution doesn't seem to have any options on how to treat boundary points.
)

**Comment by kernc** [comment]
Sorry for having it take so long. This has been updated to work with 2d data. I used and slightly amended your commit from https://github.com/statsmodels/statsmodels/issues/2820#issuecomment-185264557.
One of the tests seems to be failing, but I'm not sure whether it's my fault.

**Comment by josef-pkt** [comment]
@kernc  The test failure is not related to this PR.
I have seen it yesterday in another PR for the first time. Something must have changed on that version very recently.

**Comment by kernc** [comment]
Now that 0.8.0 is released, would you care to have another look at this? It includes your 2D patch from https://github.com/statsmodels/statsmodels/issues/2820.

**Comment by josef-pkt** [comment]
The trend extrapolation affects the estimate of the seasonal component because the detrended series is now not truncated anymore.
I think that's fine, it might create some distortion in small samples if the trend is not well estimated. (e.g. I ended up with some noisy, wiggly trend estimates where the extrapolation is sensitive to the neighboring fluctuation and transmits it to the seasonal estimate)
No problem in larger samples (I tried nobs=60) with smooth trend and extrapolation based on several, e.g. 5, observations.

One question is whether there should be an automatic bandwidth choice (number of observations) in the extrapolation trend regression, e.g. analogous to the default `filt` window length which is equal to `freq`.

**Comment by josef-pkt** [comment]
just realized 'interpolation' is a misnomer in this case, should be 'extrapolation'
(what's the keyword that scipy uses?)

**Comment by josef-pkt** [comment]
allowing `trend_interp='freq'` would allow a sensible default window, where the user doesn't need to specify the season length, if it is inferred from the date index

**Comment by josef-pkt** [comment]
a nice example

```
# 2d case
np.random.seed(987125)
nobs=60
cl = 5
x = 0.05 * np.arange(nobs)[:,None] + np.tile(np.arange(cl), (2, nobs//cl)).T  
x = x + 0.5 * np.random.randn(*x.shape)
x[:,0] += 2
decomp = seasonal_decompose(x, freq=cl, trend_interp=cl)
trend = decomp.trend
#assert_almost_equal(trend, x)
decomp.plot()

decomp2 = seasonal_decompose(x, freq=cl, trend_interp=0)
decomp2.plot()
```

trend extrapolation also looks good with nobs=10 and nobs=15
at nobs=10, only two cycles, the seasonal component is not well estimated, but looks already pretty good with nobs=15

**Comment by josef-pkt** [comment]
my suggestion: change name of keyword and allow `freq` as length
then it looks ready to merge, and works well

**Comment by kernc** [comment]
All mentions of interpolation replaced with extrapolation. I knew that. 😊 

A sensible inferred default sounds very reasonable. Added `extrapolate_trend='freq'` which takes the value of `freq`. Is that what you had in mind?

**Comment by josef-pkt** [comment]
Yes, That looks good, and is ready for merge

Can you rebase?  Not strictly necessary but makes a nicer history.

**Comment by kernc** [comment]
Already did. The branch splits from f11930db6.

**Comment by josef-pkt** [comment]
merged

@kernc  Thanks for the improvements and I'm sorry for the long delay.

**Pull request merged**